### PR TITLE
Make Preset Deletion and Creation Undoable

### PR DESCRIPTION
**Pull request by holyjewsus**
### Purpose

To make Preset creation and deletion undoable/ redoable

This PR modifies presetModel so that it inherits from **ModelBase**, this unifies presets with other models, but requires making some properties non-readonly and changing the constructor/serialization logic.

This also resulted in the removal of static constructors for presets, instead presets are created by making an instance and deserializing, like other models.

I have tried to move most of the undo/redo logic out of presets implementation and into the command framework so that calling the preset API on workspace does not inadvertently add action groups to the **undoRecorder**.

When Undo/Redo creates a new preset also raise property change event that is tied to UI visibility of presets so that they show in UI without user needing to click the view. This event bubbles from **workspaceModel** to **DynamoModel** to **DynamoViewModel**. (an alternative is to implement a **RecordableCommand** for adding a deserialized preset where we can raise the event directly on the **DynamoViewModel**)

some tests have been added for testing undo redo of preset deletion and creation and I have refactored out repeated setup steps.

- [x] will run self serve CI

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate

### Reviewers

@pboyer 
@ramramps 

### FYIs

@jnealb added tests for multiple undo/redo commands

### CODE REVIEW

File: src/DynamoCore/Models/PresetModel.cs [hunk 1]
```diff
@@ -155,61 +118,58 @@ protected virtual void SerializeCore(System.Xml.XmlElement element, SaveContext
...
-            return loadFromXmlCore(element, nodesInNodeGraph, logger);
-        }
+            var stateName = nodeElement.GetAttribute(NameAttributeName);
+            var stateguidString = nodeElement.GetAttribute(GuidAttributeName);
+            var stateDescription = nodeElement.GetAttribute(DescriptionAttributeName);
 
+            Guid stateID;
+            if (!Guid.TryParse(stateguidString, out stateID))
+            {
+                this.Log("unable to parse the GUID for preset state: " + stateName + ", will atttempt to load this state anyway");
```
Comment (ramramps) [inline src/DynamoCore/Models/PresetModel.cs]: Why logger is not used here?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 2]
```diff
@@ -840,8 +840,13 @@ void _model_PropertyChanged(object sender, PropertyChangedEventArgs e)
                         this.PublishCurrentWorkspaceCommand.RaiseCanExecuteChanged();
                     RaisePropertyChanged("IsPanning");
                     RaisePropertyChanged("IsOrbiting");
+                    RaisePropertyChanged("EnablePresetOptions");
```
Comment (ramramps) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Do we need EnablePresetOptions in both places?

File: test/DynamoCoreTests/PresetsTests.cs [hunk 3]
```diff
@@ -33,56 +31,71 @@ public void CanAddPresetState()
...
+            model.ExecuteCommand(new DynamoModel.AddPresetCommand("state1", "3", IDS));
+
+            //change values
+            numberNode1.Value = "2";
+            numberNode2.Value = "3";
+
+            DynamoSelection.Instance.ClearSelection();
+            DynamoSelection.Instance.Selection.Add(numberNode1);
+            DynamoSelection.Instance.Selection.Add(numberNode2);
+            IDS = DynamoSelection.Instance.Selection.OfType<NodeModel>().Select(x => x.GUID).ToList();
```
Comment (ramramps) [inline test/DynamoCoreTests/PresetsTests.cs]: IDS = ids.

File: test/DynamoCoreWpfTests/PresetViewModelTest.cs [hunk 4]
```diff
@@ -58,5 +58,50 @@ public void TogglePresetOptionVisibility()
...
+
+            //verify the node was created
+            Assert.AreEqual(1, ViewModel.Model.CurrentWorkspace.Nodes.Count());
+
+            //Check the Preset option visibility.
+            Assert.AreEqual(false, ViewModel.EnablePresetOptions);
+
+            DynamoSelection.Instance.Selection.Add(addNode);
+
+            var IDS = DynamoSelection.Instance.Selection.OfType<NodeModel>().Select(x => x.GUID).ToList();
```
Comment (ramramps) [inline test/DynamoCoreWpfTests/PresetViewModelTest.cs]: IDS = ids.

File: src/DynamoCore/Models/PresetModel.cs [hunk 5]
```diff
@@ -98,56 +83,34 @@ public PresetModel(string name, string description, IEnumerable<NodeModel> input
...
-        /// <returns></returns>
-        public XmlElement Serialize(XmlDocument xmlDocument, SaveContext context)
+        /// <param name="nodesInGraph"></param>
+ 
+        public PresetModel(IEnumerable<NodeModel> nodesInGraph)
         {
-            var element = CreateElement(xmlDocument, context);
-            SerializeCore(element, context);
-            return element;
+            this.nodesInWorkspaceAtConstruction = nodesInGraph.ToList();
```
Comment (ramramps) [inline src/DynamoCore/Models/PresetModel.cs]: I think ToList is not required here.

File: src/DynamoCore/Models/PresetModel.cs [hunk 6]
```diff
@@ -98,56 +81,34 @@ public PresetModel(string name, string description, IEnumerable<NodeModel> input
...
-        /// <returns></returns>
-        public XmlElement Serialize(XmlDocument xmlDocument, SaveContext context)
+        /// <param name="nodesInGraph"></param>
+ 
+        public PresetModel(IEnumerable<NodeModel> nodesInGraph)
         {
-            var element = CreateElement(xmlDocument, context);
-            SerializeCore(element, context);
-            return element;
+            this.nodes = nodesInGraph.ToList();
```
Comment (holyjewsus) [inline src/DynamoCore/Models/PresetModel.cs]: at construction time when loading a preset store all the nodes in the current node graph in the nodes field, then when it's deserialized we replace this with the nodes that the preset stores.

File: src/DynamoCore/Models/PresetModel.cs [hunk 7]
```diff
@@ -155,61 +116,59 @@ protected virtual void SerializeCore(System.Xml.XmlElement element, SaveContext
...
+                }
+            }
 
+            this.Name = stateName;
+            this.GUID = stateID;
+            this.Description = stateDescription;
+            this.serializedNodes = deserialzedNodes;
+            //at the time of deserialization, nodes contains all the nodes in the nodegraph
+            //we now replace it with the found nodes that this preset serializes
+            this.nodes = foundNodes;
```
Comment (holyjewsus) [inline src/DynamoCore/Models/PresetModel.cs]: here is where we replace the nodes field with the updated nodes we found

**Comment by holyjewsus** [comment]
@ramramps made the changes and a small change to how deserialization works so that we don't keep a list of all the nodes in the graph in each preset.

**Pull request merged**